modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +162,88 @@ impl VulnerabilityService {
 
         Ok(result.rows_affected)
     }
+
+    pub async fn analyze_purls<C: ConnectionTrait>(
+        &self,
+        purls: Vec<&str>,
+        connection: &C,
+    ) -> Result<HashMap<String, Vec<VulnerabilityDetails>>, Error> {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let query = purls.iter().filter_map(|p| {
+          let purl = Purl::from_str(p).map_err(Error::Purl);
+          let canonical = match purl {
+            Ok(purl) =>{
+              CanonicalPurl::from(purl)
+            },
+            Err(_) => return None
+          };
+
+          let version =  match canonical.version.as_ref() {
+            Some(v) => {
+              v
+            }
+            None => return None
+          };
+
+          let name = canonical.name;
+          let ty = canonical.ty;
+          let ns_condition = match canonical.namespace.clone() {
+            Some(namespace) => format!(r#"base_purl.namespace = '{namespace}'"#),
+            None => String::from("base_purl.namespace is null")
+          };
+
+          Some(format!(r#"
+            SELECT '{p}' as requested_purl,
+              vulnerability.id, vulnerability.title, vulnerability.reserved,
+              vulnerability.published, vulnerability.modified, vulnerability.withdrawn, vulnerability.cwes
+            FROM base_purl
+              LEFT JOIN purl_status ON base_purl.id = purl_status.base_purl_id
+              INNER JOIN version_range ON purl_status.version_range_id = version_range.id
+              LEFT JOIN vulnerability ON purl_status.vulnerability_id = vulnerability.id
+              INNER JOIN status ON purl_status.status_id = status.id
+            WHERE {ns_condition}
+              AND base_purl.name = '{name}'
+              AND base_purl.type = '{ty}'
+              AND version_matches('{version}', version_range.*) = TRUE
+              AND status.slug != 'fixed'"#))
+        }).collect::<Vec<String>>()
+        .join(" UNION ALL ");
```

Comment:
I really don't want to make a habit of raw sql in our services. Feels like php! :smile:

Reply:
I wasn't able to make it work using SeaORM. I could use some help here

Reply:
Especially for more complex queries, I'd prefer having some raw SQL (with proper parameters) over some hard to understanding sea_orm chained calls. It also makes it easier to debug the SQL, or work with it in other tools.